pkg/bpf: Use the other directory when /sys/fs/bpf is not BPFFS

vadorovsky opened the pull request:
When Cilium is running inside container, it has /sys/fs/bpf
directory mounted from the host. However, if BPPFS is not mounted
on the host, then containerized Cilium should be aware of that and
try to mount BPPFS in the other directory - /run/cilium/bpffs.

Fixes #3717

Signed-off-by: Michal Rostecki <mrostecki@suse.com>

```release-note
Remount /sys/fs/bpf if its filesystem is not BPFFS
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/3786)
<!-- Reviewable:end -->

vadorovsky committed:
pkg/bpf: Use the other directory when /sys/fs/bpf is not BPFFS

When Cilium is running inside container, it has /sys/fs/bpf
directory mounted from the host. However, if BPPFS is not mounted
on the host, then containerized Cilium should be aware of that and
try to mount BPPFS in the other directory - /run/cilium/bpffs.

Fixes #3717

Signed-off-by: Michal Rostecki <mrostecki@suse.com>

borkmann reviewed (commented):
@mrostecki Did you test that `/run/cilium/bpffs` gets populated by tc? tc has an option to pass the mount point via env, but it also is able to auto-detect a BPF fs mount from /proc/mounts; just making sure the latter works fine in this setting. What happens if you have, say, another BPF fs instance in some different location like /tmp/foo. Are the maps/progs still populated at the right location under /run/cilium/bpffs when /tmp/foo is listed first or last?

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+	}
+
+	log.WithField(logfields.BPFMapName, name).Debug("bpffs is not mounted yet; adding to list of maps to remove once it is mounted")
+	delayedRemoves = append(delayedRemoves, name)
```
What's is the `delayedRemoves`?

vadorovsky replied:
Delayed removes is the list of deprecated BPF maps which are going to be removed after BPFFS is mounted. I made it specially for https://github.com/cilium/cilium/blob/master/pkg/maps/tunnel/tunnel.go#L62-L63

vadorovsky replied:
TBH, I'm almost sure that the `os.Remove` call which is currently used in master is ineffective, because when `init` functions are executed when Cilium is started, BPFFS is not mounted yet. So it's de facto an another small bugfix. ;)

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
-			exec.Command("wc", "-l"),
-			exec.Command("cut", "-f1", "-d "),
+		if os.IsNotExist(err) {
+			os.MkdirAll(mapRoot, 0644)
```
If `mapRoot` is a directory it should have the execution permission bits set.

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+		if os.IsNotExist(err) {
+			os.MkdirAll(mapRoot, 0644)
+		} else {
+			return fmt.Errorf("failed to stat the mount path %s: %v", mapRoot, err)
```
`%s: %v", mapRoot, err)` to `%s: %s", mapRoot, err)`

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
-			return fmt.Errorf("command execution failed: %s", stderr)
-		}
+	if !mapRootStat.IsDir() {
+		return fmt.Errorf("%s is a file which is not a directory: %v", mapRoot, err)
```
err is nil at this point you can remove it from the `fmt.Errorf`

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
-		// Strip the newline character at the end.
-		parts := strings.Split(string(output), "\n")
+	if err := syscall.Mount(mapRoot, mapRoot, "bpf", 0, ""); err != nil {
+		return fmt.Errorf("failed to mount %s: %v", mapRoot, err)
```
`%s: %v", mapRoot, err)` to `%s: %s", mapRoot, err)`

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+func newMountsScanner() (*mountsScanner, error) {
+	fMounts, err := os.Open("/proc/mounts")
+	if err != nil {
+		return nil, fmt.Errorf("failed to open /proc/mounts: %v", err)
```
`s: %v", err)` to `s: %s", err)`

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+	return ms.scanner.Text()
+}
+
+func newMountsScanner() (*mountsScanner, error) {
```
Can you pass the filename as an argument to this function so you can add a unit test?

```
// search and replace newMountsScanner() with mountScanner()
func mountsScanner() (*mountsScanner, error) {
   return newMountsScanner("/proc/mounts")
}

// unit test this function
func newMountsScanner(path string) (*mountsScanner, error) {
....
}
```

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+// isBPFFS checks whether the current mapRoot:
+// - has any mount
+// - if yes, it checks whether the filesystem of that mount is BPFFS
+func isBPFFS() (bool, bool, error) {
```
in the function description, write what are the returned values

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+//    probably Cilium is running inside container which has mounted /sys/fs/bpf
+//    from host, but host doesn't have proper BPFFS mount, so that mount is just
+//    the empty directory. In that case, mount BPFFS under /run/cilium/bpffs.
+func checkOrMountDefaultLocations() error {
```
The function description is not the same as the function's name

vadorovsky replied:
done

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+		log.Warnf("BPF filesystem is going to be mounted automatically in %s. However, it probably "+
+			"means that Cilium is running inside container and BPFFS is not mounted on "+
+			"the host. We recommend to do that: "+
+			"https://cilium.readthedocs.io/en/latest/kubernetes/install/?highlight=bpffs#mounting-the-bpf-fs-optional",
```
This probably needs a short URL similar the one with have for the kvstore error cc @tgraf

vadorovsky replied:
What kind of short URL do you expect? `git grep "http\:\/\/" pkg/kvstore/` gave me only:

```
pkg/kvstore/allocator/allocator.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/allocator/allocator_test.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/allocator/doc.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/allocator/localkeys.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/allocator/localkeys_test.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/allocator/logfields.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/backend.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/backwards_compat.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/base_test.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/client.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/config.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/consul.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/consul_test.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/doc.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/dummy.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/etcd.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/etcd.go:    etcdDummyAddress = "http://127.0.0.1:4002"
pkg/kvstore/etcd_test.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/etcd_test.go:                       case "http://127.0.0.1:4004":
pkg/kvstore/etcd_test.go:       v, err := getEPVersion(mm, "http://127.0.0.1:4003", time.Second)
pkg/kvstore/etcd_test.go:       v, err = getEPVersion(mm, "http://127.0.0.1:4004", time.Second)
pkg/kvstore/etcd_test.go:       cfg.Endpoints = []string{"http://127.0.0.1:4003", "http://127.0.0.1:4005"}
pkg/kvstore/etcd_test.go:       cfg.Endpoints = []string{"http://127.0.0.1:4003", "http://127.0.0.1:4004", "http://127.0.0.1:4005"}
pkg/kvstore/events.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/keepalive.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/kvstore.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/kvstore_test.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/lock.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/logfields.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/store/doc.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/store/store.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/store/store_test.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/trace.go://     http://www.apache.org/licenses/LICENSE-2.0
pkg/kvstore/watcher_cache.go://     http://www.apache.org/licenses/LICENSE-2.0
```

aanm commented on pkg/bpf/bpffs.go:
```diff
@@ -104,100 +118,275 @@ func OpenAfterMount(m *Map) error {
+		}
+	}
+
+	multipleMounts, err := hasMultipleMounts()
```
If you ware checking if the filepath has multiple mounts **after** mounting, doesn't this means cilium will keep failing afterwards?

vadorovsky replied:
No. After mounting Cilium should have only one mount (`hasMultipleMounts` returns true if number of mounts is greater than 1).

vadorovsky committed:
common/files: Add fileScanner struct

Add fileScanner struct which wraps the file descriptor and provides
a bufio.Scanner object to read the content of file.

Signed-off-by: Michal Rostecki <mrostecki@suse.com>
